### DISCUSSION TITLE
Two ways to get explanation (important edges) of a trained GNN model

### DISCUSSION THREAD

**Discussion by Wienhannover** [pyg-team/pytorch_geometric#7083] [Q&A]
Hi, when it comes to explaining a trained GNN model, is there any difference between using learnable edge weights, such as
#------------------------------------------------------------------------------
masked_prediction = model_to_explain(feats, graph, edge_weights=self.edge_mask)
#------------------------------------------------------------------------------
and
#------------------------------------------------------------------------------
self.edge_mask = torch.nn.Parameter(torch.randn(E) * std)
for module in self.model_to_explain.modules():
****if isinstance(module, MessagePassing):
********module.explain = True
********module.edge_mask = self.edge_mask
masked_prediction = model_to_explain(feats, graph)
#------------------------------------------------------------------------------
?

**Comment by rusty1s** [comment]
It depends on the underlying explainer algorithm and model. For example, not all PyG models support `edge_weight` as input argument, but explainers do not have this restriction. For `GNNExplainer`, you also want to incorporate regularization terms to let the model learn to only maintain important edges. Lastly, this approach limits you to gradient-based explainers.

**Reply by Wienhannover** [reply]
thanks for your reply! In the case that edge_weight can be applied to models, can we think that these two methods are equal, which means that both methods can output same important edges(if we choose the top-10 edges)?

**Reply by rusty1s** [reply]
Yes.

**Reply by Wienhannover** [reply]
Hi, I have a follow-up question, if I mannually set all edge weights to zero, I will still have a prediciton, in this case, how can we think of the prediction? Can we think that the prediction as the bias that is encoded in the trained model?

**Reply by rusty1s** [reply]
If all your edge weights are zero, that would mean no information is exchanged between nodes in message passing, and so the prediction would be solely contributed to the features of the source node. Hope that makes sense.

**Reply by Wienhannover** [reply]
thanks, I really appreaciate that you always reply in such a short time.:)
